src/core/lwt_stream.ml

```diff
@@ -130,28 +132,31 @@ let clone s =
      | _ -> ());
   {
     source = s.source;
+    closed = s.closed;
     node = s.node;
     last = s.last;
     hooks = s.hooks;
   }
 
-let from f =
+let from_source source =
   let last = new_node () in
-  {
-    source = From { from_create = f; from_thread = Lwt.return_unit };
-    node = last;
-    last = ref last;
-    hooks = ref [];
+  let closed, closed_w = Lwt.wait () in
+  let mark_closed () =
+    if not (Lwt.is_sleeping closed) then Lwt.wakeup closed_w () in
```

Comment:
It would be clearer to explicitly wakeup the `closed` thread at the end of the stream rather than rely on the execution of this hook. Plus with the current code, I think you get `Lwt_stream.is_closed stream = false` inside `on_termination` callbacks given that this hook is executed last. Maybe that was the indented behavior? In any case it'd be good to have tests for this in `tests/core/test_lwt_stream.ml`.

If you want to avoid storing both the `Lwt.t` and the `Lwt.u` inside the `Lst_stream.t`, you can store only the latter and define `closed` as follow, which is a no-op:

```ocaml
let closed s = Lwt.waiter_of_wakener s.closed
```